File: javascript/game.js
exports = typeof window !== "undefined" && window !== null ? window : global;

exports.Game = function() {

  const maxNumberOfPlayers = 6;

  var players          = new Array();
  var places           = new Array(maxNumberOfPlayers);
  var purses           = new Array(maxNumberOfPlayers);
  var inPenaltyBox     = new Array(maxNumberOfPlayers);

  var popQuestions     = new Array();
  var scienceQuestions = new Array();
  var sportsQuestions  = new Array();
  var rockQuestions    = new Array();

  var currentPlayer    = 0;
  var isGettingOutOfPenaltyBox = false;

  var didPlayerWin = function(){
    return !(purses[currentPlayer] == maxNumberOfPlayers)
  };

  var currentCategory = function(){
    if(places[currentPlayer] == 0)
      return 'Pop';
    if(places[currentPlayer] == 4)
      return 'Pop';
    if(places[currentPlayer] == 8)
      return 'Pop';
    if(places[currentPlayer] == 1)
      return 'Science';
    if(places[currentPlayer] == 5)
      return 'Science';
    if(places[currentPlayer] == 9)
      return 'Science';
    if(places[currentPlayer] == 2)
      return 'Sports';
    if(places[currentPlayer] == 6)
      return 'Sports';
    if(places[currentPlayer] == 10)
      return 'Sports';
    return 'Rock';
  };

  this.createRockQuestion = function(index){
    return "Rock Question "+index;
  };

  for(var i = 0; i < 50; i++){
    popQuestions.push("Pop Question "+i);
    scienceQuestions.push("Science Question "+i);
    sportsQuestions.push("Sports Question "+i);
    rockQuestions.push(this.createRockQuestion(i));
  };

  this.isPlayable = function(){
    return players.length >= 2;
  };

  this.add = function(playerName){
      console.log(`Adding ${playerName} current number of players is ${this.howManyPlayers()}`)
      if (this.canAddNewPlayer(this.howManyPlayers())) {
          players.push(playerName);
          places[this.howManyPlayers() - 1] = 0;
          purses[this.howManyPlayers() - 1] = 0;
          inPenaltyBox[this.howManyPlayers() - 1] = false;

          console.log(playerName + " was added");
          console.log("They're player number " + players.length);

          return true;
      } else {
          console.log(`There are ${players.length} in this game, the max is 6`)
          console.log(`Couldn't add ${playerName}, There are too many players in this game!`);
          return false
      }
  };

  this.getCurrentPlayer = function() {
    return players[currentPlayer];
  };

  this.howManyPlayers = function(){
    return players.length;
  };

  this.canAddNewPlayer = function (currentCount) {
      return currentCount < maxNumberOfPlayers;
  }

  var askQuestion = function(){
    if(currentCategory() == 'Pop')
      console.log(popQuestions.shift());
    if(currentCategory() == 'Science')
      console.log(scienceQuestions.shift());
    if(currentCategory() == 'Sports')
      console.log(sportsQuestions.shift());
    if(currentCategory() == 'Rock')
      console.log(rockQuestions.shift());
  };

  /*
  * playRound()
  * responsible for:
  * - rolling dice for current player
  * - asking the question
  * - accepting the answer
  * - determining if there is a winner \ has the game ended?
  * -- no? setting the next player
  * -- yes? end the game
  *
  * */

  this.roll = function(roll){

    if (this.isPlayable()) {
      console.log(players[currentPlayer] + " is the current player");
      console.log("They have rolled a " + roll);

      if(inPenaltyBox[currentPlayer]){
        if(roll % 2 != 0){
          isGettingOutOfPenaltyBox = true;

          console.log(players[currentPlayer] + " is getting out of the penalty box");
          places[currentPlayer] = places[currentPlayer] + roll;
          if(places[currentPlayer] > 11){
            places[currentPlayer] = places[currentPlayer] - 12;
          }

          console.log(players[currentPlayer] + "'s new location is " + places[currentPlayer]);
          console.log("The category is " + currentCategory());
          askQuestion();
        }else{
          console.log(players[currentPlayer] + " is not getting out of the penalty box");
          isGettingOutOfPenaltyBox = false;
        }
      }else{

        places[currentPlayer] = places[currentPlayer] + roll;
        if(places[currentPlayer] > 11){
          places[currentPlayer] = places[currentPlayer] - 12;
        }

        console.log(players[currentPlayer] + "'s new location is " + places[currentPlayer]);
        console.log("The category is " + currentCategory());
        askQuestion();
      }
      return true;
    } else {
      return false;
    }

  };

  this.wasCorrectlyAnswered = function(){
    if(inPenaltyBox[currentPlayer]){
      if(isGettingOutOfPenaltyBox){
        console.log('Answer was correct!!!!');
        purses[currentPlayer] += 1;
        console.log(players[currentPlayer] + " now has " +
                    purses[currentPlayer]  + " Gold Coins.");

        var winner = didPlayerWin();
        currentPlayer += 1;
        if(currentPlayer == players.length)
          currentPlayer = 0;

        return winner;
      }else{
        currentPlayer += 1;
        if(currentPlayer == players.length)
          currentPlayer = 0;
        return true;
      }



    }else{

      console.log("Answer was correct!!!!");

      purses[currentPlayer] += 1;
      console.log(players[currentPlayer] + " now has " +
                  purses[currentPlayer]  + " Gold Coins.");

      var winner = didPlayerWin();

      currentPlayer += 1;
      if(currentPlayer == players.length)
        currentPlayer = 0;

      return winner;
    }
  };

  this.wrongAnswer = function(){
		console.log('Question was incorrectly answered');
		console.log(players[currentPlayer] + " was sent to the penalty box");
		inPenaltyBox[currentPlayer] = true;

    currentPlayer += 1;
    if(currentPlayer == players.length)
      currentPlayer = 0;
		return true;
  };
};
/*var notAWinner = false;

var game = new Game();

game.add('Chet');
game.add('Pat');
game.add('Sue');

do{

  game.playRound();

  game.roll(Math.floor(Math.random()*6) + 1);

  if(Math.floor(Math.random()*10) == 7){
    notAWinner = game.wrongAnswer();
  }else{
    notAWinner = game.wasCorrectlyAnswered();
  }

}while(!game.isThereAWinner());*/

File: rust/Game.rs
pub struct Game
{
    players        : Vec<String>,
    places         : [i32; 6],
    purses         : [i32; 6],
    in_penaltybox  : [bool; 6],
    current_player : i32,
    is_getting_out_of_penaltybox : bool,

    pop_questions     : Vec<String>,
    science_questions : Vec<String>,
    sports_questions  : Vec<String>,
    rock_questions    : Vec<String>,
}

impl Default for Game {
    fn default() -> Game {
        let mut game = Game {
            players        : vec![],
            places         : [0;6],
            purses         : [0;6],
            in_penaltybox  : [false;6],
            current_player : 0,
            is_getting_out_of_penaltybox : false,
            pop_questions     : Vec::new(),
            science_questions : Vec::new(),
            sports_questions  : Vec::new(),
            rock_questions    : Vec::new(),
        };
        for x in 0..50 {
            let pop_qu = "Pop Question ".to_string() + &x.to_string();
            game.pop_questions.push(pop_qu);
            let sci_qu = "Science Question ".to_string() + &x.to_string();
            game.science_questions.push(sci_qu);
            let spo_qu = "Sports Question ".to_string() + &x.to_string();
            game.sports_questions.push(spo_qu);
            let mut rock_qu = game.create_rock_question(x);
            game.rock_questions.push(rock_qu);
        }
        game
    }
}

impl Game {
    fn how_many_players(&self) -> usize {
        self.players.len()
    }
    fn is_playable(&self) -> bool {
        self.how_many_players() >= 2
    }
    fn did_player_win(&self) -> bool {
        !(self.purses[self.current_player as usize] == 6)
    }
    fn current_category(&self) -> &'static str {
        match self.places[self.current_player as usize] {
            0 | 4 | 8  => return "Pop",
            1 | 5 | 9  => return "Science",
            2 | 6 | 10 => return "Sports",
            _          => return "Rock"
        }
    }
    fn create_rock_question(&self,index: i32) -> String {
        "Rock Question ".to_string() + &index.to_string()
    }
}

impl Game {
    fn add(&mut self,player_name: String) -> bool {
        let l_player = player_name.clone();
        self.players.push(player_name);
        self.places[self.how_many_players()] = 0;
        self.purses[self.how_many_players()] = 0;
        self.in_penaltybox[self.how_many_players()] = false;
        println!("{} was added",l_player);
        println!("They are player number {}",self.players.len());
        true
    }
}

impl Game {
    fn wrongAnswer(&mut self) -> bool {
        println!("Question was incorrectly answered");
        println!("{} was sent to the penalty box",self.players[self.current_player as usize]);
        self.in_penaltybox[self.current_player as usize] = true;
        self.current_player += 1;
        if(self.current_player == self.players.len() as i32) {
            self.current_player = 0;
        }
        true
    }
}

fn main()
{
    let mut game = Game {..Default::default()}; 
    game.add("Player2".to_string());
}
